axelor-base/src/main/java/com/axelor/apps/base/service/ProductCompanyServiceImpl.java

```diff
@@ -99,4 +113,49 @@ private Product findAppropriateProductCompany(
 
     return product;
   }
+
+  /**
+   * Finds the appropriate ONLY company-specific version of a product if searched field is
+   * overwritten by company
+   *
+   * @param originalProduct
+   * @param fieldName
+   * @param company
+   * @return
+   * @throws AxelorException
+   */
+  protected Product findAppropriateProductCompanyWithNoDefault(
+      Product originalProduct, String fieldName, Company company) throws AxelorException {
+    if (originalProduct == null) {
+      throw new AxelorException(
+          TraceBackRepository.CATEGORY_MISSING_FIELD,
+          I18n.get(BaseExceptionMessage.PRODUCT_COMPANY_NO_PRODUCT),
+          fieldName);
+    } else if (fieldName == null || fieldName.trim().equals("")) {
+      throw new AxelorException(
+          TraceBackRepository.CATEGORY_MISSING_FIELD,
+          I18n.get(BaseExceptionMessage.PRODUCT_COMPANY_NO_FIELD),
+          originalProduct.getFullName());
+    }
+
+    Product product = null;
+
+    if (company != null && originalProduct.getProductCompanyList() != null) {
+      for (ProductCompany productCompany : originalProduct.getProductCompanyList()) {
+        if (company.equals(productCompany.getCompany())) {
+          Set<MetaField> companySpecificFields =
+              appBaseService.getAppBase().getCompanySpecificProductFieldsSet();
+          for (MetaField field : companySpecificFields) {
```

Comment:
Why do you need to do this check inside the loop ? fieldName is available at the start of the method. Please rewrite this method as it is not optimized at all and is difficult to understand